### PR TITLE
[ARM] Fix infinite recursion in unwind

### LINKED ISSUE #119905: Inductor test segfault on macos (and arm64) when C++ stacktraces are enabled
### 🐛 Describe the bug

This was discovered while working on https://github.com/pytorch/pytorch/pull/119408 that tries to automatically enable TORCH_SHOW_CPP_STACKTRACES for the re-runs of failing jobs.

On macos, this consistently leads to test files in test/inductor/ to segfault.

An example is https://github.com/pytorch/pytorch/actions/runs/7823402867/job/21344456540 and you can find other in the CI of the PR.
Sending a PR that removes the special case for inductor should trigger this again as well.

### Versions

Main

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### PR DESCRIPTION
**Pull request by Aidyn-A**
Fixes #119905

The `TORCH_SHOW_CPP_STACKTRACES=1` setting on ARM causes infinite recursive unwind because on failure a `StackTraceFetcher` attempts to unwind the <ins>failed instruction</ins>: https://github.com/pytorch/pytorch/blob/5ad759ca33ba8299cf7e1a6bb1dff7c9a5555e29/torch/csrc/profiler/combined_traceback.cpp#L25
then the unwind itself fails:
https://github.com/pytorch/pytorch/blob/5ad759ca33ba8299cf7e1a6bb1dff7c9a5555e29/torch/csrc/profiler/unwind/unwind.cpp#L10-L12
and it causes another attempt to unwind the failure in `unwind()`...

In summary, the executed instruction is equivalent to:
```C++
std::vector<void*> unwind() {
  // some instructions ...
  return unwind();
}
```
This PR replaces `TORCH_CHECK` by `TORCH_WARN_ONCE` as it will not cause an uncontrolled recursion. The only side effect would be an empty back-trace. 

Huge thanks to @nWEIdia who found the root cause!

cc @malfet @snadampal @milpuz01

### CODE REVIEW

**Review by eqy — APPROVED** [review]

**Review by nWEIdia — APPROVED** [review]
Geniously speed of light forward fix!

**Review by malfet — APPROVED** [review]

**Comment by nWEIdia** [comment]
Perhaps as a follow up, we can remove [IS_MACOS ](https://github.com/pytorch/pytorch/blob/main/test/run_test.py#L29) and the [inductor check line](https://github.com/pytorch/pytorch/blob/main/test/run_test.py#L579)?

File: test/run_test.py [hunk 1]
```diff
@@ -573,11 +573,7 @@ def run_test(
 
 def try_set_cpp_stack_traces(env, command, set=True):
     # Print full c++ stack traces during retries
-    # Don't do it for macos inductor tests as it makes them
-    # segfault for some reason
-    if not (
-        IS_MACOS and len(command) >= 2 and command[2].startswith(INDUCTOR_TEST_PREFIX)
-    ):
+    if not (len(command) >= 2 and command[2].startswith(INDUCTOR_TEST_PREFIX)):
```
Comment (nWEIdia) [inline test/run_test.py:576]: The above would disable torch inductor print, on both MACOS and ARM. 

The original intention was to disable MACOS's torch inductor tests' TORCH_SHOW_CPP_STACKTRACES. 

We would to enable them, so we could just delete the entire if command. 

Stacktrace would be printed across the board, and should have no issues. 

BTW, thank you for addressing in this PR. I meant to say in a follow-up PR.

Comment (nWEIdia) [reply test/run_test.py:576]: The import of IS_MACOS can be deleted too as this is the only occurence.

File: test/run_test.py [hunk 2]
```diff
@@ -573,9 +573,8 @@ def run_test(
 
 def try_set_cpp_stack_traces(env, command, set=True):
     # Print full c++ stack traces during retries
-    if not (len(command) >= 2 and command[2].startswith(INDUCTOR_TEST_PREFIX)):
-        env = env or {}
-        env["TORCH_SHOW_CPP_STACKTRACES"] = "1" if set else "0"
+    env = env or {}
+    env["TORCH_SHOW_CPP_STACKTRACES"] = "1" if set else "0"
```
Comment (nWEIdia) [inline test/run_test.py:577]: Nit, Don't forget to remove the original import IS_MACOS at the beginning section of the file :)

File: test/run_test.py (hunk 1 above)
Comment (Aidyn-A) [reply test/run_test.py]: >The import of IS_MACOS can be deleted too as this is the only occurrence.

Nope, it is not the only occurrence:
https://github.com/pytorch/pytorch/blob/fc61aae70f249c9b601fac6837df0e3ccef4f71f/test/run_test.py#L1268-L1289

Comment (nWEIdia) [reply test/run_test.py]: You are right!

**Comment by Aidyn-A** [comment]
All tests are green.

@pytorchbot merge
